### DISCUSSION TITLE
convert ed25519 key to x25519

### DISCUSSION THREAD

**Discussion by xicilion** [openssl/openssl#23905] [Q&A] [answered]
In many applications we need to convert x25519 key from ed25519 key for DH exchange, at present I don't seem to find this function in openssl public api, and the public api of curve25519 in openssl is not enough to realize this conversion.

In order to support this conversion, I modified src/crypto/ec/curve25519.c and added the following two functions:

```cpp
void
ossl_x25519_private_from_ed25519(uint8_t out_private_key[32],
                                 const uint8_t private_key[32])
{
    uint8_t md[SHA512_DIGEST_LENGTH];

    SHA512(private_key, 32, md);

    md[0] &= 248;
    md[31] &= 127;
    md[31] |= 64;

    memcpy(out_private_key, md, 32);

    OPENSSL_cleanse(md, sizeof(md));
}

void
ossl_x25519_public_from_ed25519(uint8_t out_public_value[32],
                                const uint8_t public_value[32])
{
    fe Z, Y;
    fe zplusy, zminusy, zminusy_inv;

    fe_frombytes(Y, public_value);
    fe_1(Z);

    fe_add(zplusy, Z, Y);
    fe_sub(zminusy, Z, Y);
    fe_invert(zminusy_inv, zminusy);
    fe_mul(zplusy, zplusy, zminusy_inv);

    fe_tobytes(out_public_value, zplusy);
}
```
This is a function that currently has no particular claim to use in openssl, so I hope it's useful to you.

**Comment by nhorman** [comment, marked as answer]
I think the proscribed method to do what you want to do is to use the EVP_PKEY_fromdata api call to export the keypair information from the ED25519 key, and then use those params to import the data to a X25519 key using EVP_PKEY_todata

**Comment by xicilion** [comment]
These two functions only realize the conversion of ed25519 public-private key to X25519 public-private key. In the end, when you use it, you do need to use the method you mentioned, take the raw data from the PKey, do the calculation, and then set the PKey of X25519 as much as possible.
The reason I added functions to src/crypto/ec/curve25519.c is that I don't want to introduce a new library of ed25519 base calculation functions.

**Comment by t8m** [comment]
I do not actually think that supporting this conversion in the library is a good idea. Using the same key for multiple purposes is in general bad security practice.

**Comment by xicilion** [comment]
You're right, but there are still some applications that require this conversion, so I'm just sharing it here as a patch.

**Reply by fencepost-error** [reply]
Just wanted to say thanks for this, I was doing some testing of data in a conference paper and this is exactly what I needed.  So it has been useful, even if not directly for OpenSSL :-).

**Comment by xicilion** [comment]
I proposed a pr for this patch.
https://github.com/openssl/openssl/pull/24621
